### DISCUSSION TITLE
How to build jaxlib from source with modified TensorFlow repository

### DISCUSSION THREAD

**Discussion by JiepingLLL** [jax-ml/jax#11630] [Q&A]
My environment: python3.6
Need to install jax==0.2.17; jaxlib==0.1.69
But I need to build the jaxlib myself and not install it directly with pip. Now there are two problems, one is that I downloaded bazel and added it to bashrc, but when I execute the command "python build/build.py", I get an error "ERROR: Client cwd'/../.. /jax/build' is not inside workpce '/../../jax' b' ' cannot find or download bazel. Please install bazel". This leaves me wondering what to do.
The second is that I see "modify the WORKSPACE file in the root of the JAX source tree to point to a different TensorFlow tree." How to modify the workspace content, I see that the content inside is roughly the same as the tensorflow workspace file the same. I'm so desperate to get your help, it's so important to me.

**Comment by jakevdp** [comment]
Sorry you're having trouble with this. It's hard to tell what might be happening without more information. Presumably you've found the build instructions here? https://jax.readthedocs.io/en/latest/developer.html#building-jaxlib-from-source

**Comment by JiepingLLL** [comment]
Yes, I built the jaxlib through the build instructions and after a day's effort it has been installed successfully via python build/build.py on ubuntu. But I have some doubts. After using the python build/build.py command, I see that the terminal has downloaded and used the bazel tool and loaded a lot of packages (this step has been running for a long time), about 174 packages. Because my ultimate goal is to install jaxlib offline through python3.6 on another platform, this platform cannot access the Internet, so the automatically loaded packages after this command (python build/build.py) is impossible on my destination platform. So I wonder what is the rationale behind build.py? I need to make it suitable for my platform by modifying build.py, but where do I need to find the packages it loads? Finally, I uploaded the information about how I installed jaxlib in ubuntu in the attachment. If possible, I hope you can help me take a look and answer my question, thank you very much
[Problems building jaxlib from build.py.pdf](https://github.com/google/jax/files/9165872/Problems.building.jaxlib.from.build.py.pdf)
.

**Comment by jakevdp** [comment]
The rationale behind `build.py` is that jaxlib is designed to be built using the `bazel` tool, and it has a number of external compile-time requirements, the most complicated being the XLA source which is extracted by `build.py` from the tensorflow repository.

What you're trying to do – i.e. build jaxlib for a currently unsupported Python version on a system that does not have the abilty to download build dependencies – sounds like it could be quite complicated.

**Reply by JiepingLLL** [reply]
Where can I find the XLA source for this tensorflow repository? I only found this https://github.com/tensorflow/tensorflow, but what is the specific XLA? Can I install the bazel tools first and then build from the XLA source?

**Reply by hawkinsp** [reply]
XLA is inside the TF repository (`tensorflow/compiler/xla`). You can build JAX with a different TF tree by following the instructions in the WORKSPACE file to point to a git checkout of TF (or a different commit hash).

**Reply by JiepingLLL** [reply]
Thanks for the guidance, but I'm sorry this page is not accessible (404 is displayed), is there any other way?

**Reply by JiepingLLL** [reply]
I found xla (tensorflow/compiler/xla), but in the WORKSPACE ( tensorflow/WORKSPACE ) file I just saw a few lines, I didn't understand it, can you give me some pointers?
